### DISCUSSION TITLE
Difference between pingInterval: 1000 and setInterval(1000) for recovery

### DISCUSSION THREAD

**Discussion by minchanhan** [socketio/socket.io#4993] [Q&A]
https://socket.io/docs/v4/connection-state-recovery#usage,
https://codesandbox.io/p/sandbox/github/socketio/socket.io/tree/main/examples/connection-state-recovery-example/esm?file=index.js

I'm following this doc and example to test out the connection state recovery feature in socket.io but I'm confused because I thought socket.io handled ping pong emits automatically, so I figured setting pingInterval: 1000 would replace the line used in the example:

```
setInterval(() => {
  io.emit("ping", new Date().toISOString());
}, 1000);
```
But the recovery isn't working when using:

```
const io = new Server(server, { // for work with socket.io
  cors: corsOptions,
  pingInterval: 1000,
  connectionStateRecovery: {
    maxDisconnectionDuration: 2 * 60 * 1000, // 2 mins backup
    skipMiddlewares: true,
  },
});
```
without the setInterval...

With some digging I found that `socket.on("ping")` isn't triggered from `pingInterval` like I expected. So do I still need this setInterval if I want to handle temporary reconnections rather than relying on pingInterval?:
```
const socket = io(
  process.env.REACT_APP_SERVER || "http://localhost:3001",
  {
    reconnectionDelay: 5000, // defaults to 1000
    reconnectionDelayMax: 5000, // defaults to 5000
  }
);
socket.on("connect", () => {
  console.log(socket.recovered); // true only for setInterval method
});
socket.io.on("ping", () => {
  console.log("ping from pingInterval only");
});
socket.on("ping", () => {
  console.log("ping from setInterval only");
});
```

**Comment by darrachequesne** [comment]
Hi!

> So do I still need this setInterval if I want to handle temporary reconnections rather than relying on pingInterval?:

The `io.emit()` is done in order to init the offset on the client side, so that the connection state can be successfully recovered afterwards.

`io.emit("ping")` is indeed not related to the heartbeat mechanism, we could have used any other event name like `io.emit("hello")` instead.

**Reply by minchanhan** [reply]
Hello, in that case, wouldn't it be sufficient to just do `io.emit("offset_ping")` on connection? I removed the setInterval and added this and the recovery still seems to be working.
